staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -489,26 +492,63 @@ func PollImmediateInfinite(interval time.Duration, condition ConditionFunc) erro
 func PollUntil(interval time.Duration, condition ConditionFunc, stopCh <-chan struct{}) error {
 	ctx, cancel := contextForChannel(stopCh)
 	defer cancel()
-	return WaitFor(poller(interval, 0), condition, ctx.Done())
+	return poll(ctx, false, poller(interval, 0), condition.WithContext())
 }
 
 // PollImmediateUntil tries a condition func until it returns true, an error or stopCh is closed.
 //
 // PollImmediateUntil runs the 'condition' before waiting for the interval.
 // 'condition' will always be invoked at least once.
 func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh <-chan struct{}) error {
-	done, err := condition()
-	if err != nil {
-		return err
-	}
-	if done {
-		return nil
+	ctx, cancel := contextForChannel(stopCh)
+	defer cancel()
+	return poll(ctx, true, poller(interval, 0), condition.WithContext())
+}
+
+// PollUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollUntilWithContext always waits interval before the first run of 'condition'.
+// 'condition' will always be invoked at least once.
+func PollUntilWithContext(ctx context.Context, interval time.Duration, condition ConditionWithContextFunc) error {
+	return poll(ctx, false, poller(interval, 0), condition)
+}
+
+// PollImmediateUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollImmediateUntilWithContext runs the 'condition' before waiting for the interval.
+// 'condition' will always be invoked at least once.
+func PollImmediateUntilWithContext(ctx context.Context, interval time.Duration, condition ConditionWithContextFunc) error {
+	return poll(ctx, true, poller(interval, 0), condition)
+}
+
+// Internally used, each of the the public 'Poll*' function defined in this
+// package should invoke this internal function with appropriate parameters.
+// ctx: the context specified by the caller, for infinite polling pass
+// a context that never gets cancelled or expired.
+// immediate: if true, the 'condition' will be invoked before waiting for the interval,
+// in this case 'condition' will always be invoked at least once.
+// wait: user specified WaitFunc function that controls at what interval the condition
+// function should be invoked periodically and whether it is bound by a timeout.
+// condition: user specified ConditionWithContextFunc function.
+func poll(ctx context.Context, immediate bool, wait WaitFunc, condition ConditionWithContextFunc) error {
+	if immediate {
+		done, err := condition(ctx)
```

Comment:
doing it like this, you just lost crash protection on this call, didn't you?

Reply:
This is in keeping with the current behavior - all immediate invocations of the condition function are not crash protected, I added crash protection.